metadata/graphsync_filecoinv1.go

```diff
@@ -0,0 +1,87 @@
+package metadata
+
+import (
+	"bytes"
+	"fmt"
+
+	stiapi "github.com/filecoin-project/storetheindex/api/v0"
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	"github.com/ipld/go-ipld-prime/node/bindnode"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/multiformats/go-multicodec"
+)
+
+// GraphsyncFilecoinV1Metadata represents the indexing metadata for multicodec.TransportGraphsyncFilecoinv1.
+type GraphsyncFilecoinV1Metadata struct {
+	// PieceCID identifies the piece this data can be found in
+	PieceCID cid.Cid
+	// VerifiedDeal indicates if the deal is verified
+	VerifiedDeal bool
+	// FastRetrieval indicates whether the provider claims there is an unsealed copy
+	FastRetrieval bool
+}
+
+var graphSyncfilecoinV1SchemaType schema.Type
+
+func init() {
+	ts := schema.TypeSystem{}
+	ts.Init()
+	ts.Accumulate(schema.SpawnLink("Link"))
+	ts.Accumulate(schema.SpawnBool("Bool"))
+	ts.Accumulate(schema.SpawnStruct("GraphsyncFilecoinV1Metadata",
+		[]schema.StructField{
+			schema.SpawnStructField("PieceCID", "Link", false, false),
+			schema.SpawnStructField("VerifiedDeal", "Bool", false, false),
+			schema.SpawnStructField("FastRetrieval", "Bool", false, false),
+		},
+		schema.SpawnStructRepresentationMap(nil),
+	))
+	graphSyncfilecoinV1SchemaType = ts.TypeByName("GraphsyncFilecoinV1Metadata")
+}
+
+// ToIndexerMetadata converts GraphsyncFilecoinV1Metadata into indexer Metadata format.
+func (dtm *GraphsyncFilecoinV1Metadata) ToIndexerMetadata() (stiapi.Metadata, error) {
+	nd := bindnode.Wrap(dtm, graphSyncfilecoinV1SchemaType)
+	buf := new(bytes.Buffer)
+	err := dagcbor.Encode(nd, buf)
+	if err != nil {
+		return stiapi.Metadata{}, err
+	}
+	return stiapi.Metadata{
+		ProtocolID: multicodec.TransportGraphsyncFilecoinv1,
+		Data:       buf.Bytes(),
+	}, nil
```

Comment:
Just a note, but i think we can in a protocol-compatible way, say that metadata is not simply `<protocol ID><payload>` but rather `[<protocol id><protocol-speicfic-payload>]`, so e.g.
`uvarint{graphsync filecoin} || cbor{<cid>, true, true} || uvarint{bitswap}` to indicate that both protocols are supported.
since the cbor payload of graphsync can already be determined where the end is, if there are trailing bytes they can be interpreted as the next protocol

Reply:
Captured: https://github.com/filecoin-project/index-provider/issues/188 
Leaving this out for another PR.


If I understand the comment correctly, it means the protocol ID GraphsyncFileCoinV1 can specify multiple retrieval protocols?  If so this might seem slightly odd, where putting Bitswap metadata first means the user can't specify multiple.  Is the suggestion that, basically metadata can be nested?

From API perspective I am not sure if this needs changes in the go-indexer-core.

Cc @gammazero

Reply:
Nope, it would be in either order. Probably we'd say that when multiple protocols are supported by the provider they should be listed in metadata in numerical order (so bitswap, then data transfer).

The constraint i'm noting we need for that is that payloads need to be parseable in a way where we know when they've ended. We're lucky in that our defined payloads so far do have that property all ready, so this is a backwards compatible change to the definition of metadata structure.

Reply:
Gotcha thank you